File: perl/tasser/tasser1.0/mkinput/mkdat.pl
#!/usr/bin/perl

use Getopt::Std;
use small_util qw(addAbsPath pastry $codedir);

sub message(){
    #system("clear");
    print "Usage: mkdat.pl -d -o [-b -c -l -t -w]\n";
    print " Mandatory:\n";
    print " -d data_dir (where to find seq.fasta, seq.dat, template.dat and contact.dat. Default=current dir)\n";
    print " -o output directory (def='.')\n";
    print " Optional:\n";
    print " -b bin_dir  (def=$codedir/perl/tasser/tasser1.0/mkinput/bin)\n";
    print " -c comm_dir (def=$codedir/perl/tasser/tasser1.0/mkinput/data)\n";
    print " -l libdir   (default=/library/yzhang/PDB)\n";
    print " -t database (default=/library/yzhang/nr/nr.filter)\n";
    print " -w work_dir (default=./tmp_mkdat)\n";
    print " -v sequence file in FASTA format, located in data_dir (def='seq.fasta')\n";
    print " -x secondary structure propensity, located in data_dir (def='seq.dat')\n";
    print " -y template file, located in data_dir (def='template.dat')\n";
    print " -z predicted contacts, located in data_dir (def='contact.dat')\n";
    print "\n";
    exit(0);
}

sub parse_input(){
    my $f=0;
    if(!getopts('b:c:d:l:w:t:v:x:y:z:o:h')){ &message(); }
    if($opt_d){ $data_dir=addAbsPath($opt_d); $f++; }
    if($opt_b){ $bin_dir =addAbsPath($opt_b);       }
    if($opt_c){ $comm_dir=addAbsPath($opt_c);       }
    if($opt_l){ $libdir  =addAbsPath($opt_l);       }
    if($opt_w){ $work_dir=addAbsPath($opt_w);       }
    if($opt_t){ $db      =addAbsPath($opt_t);       }
    if($opt_v){ $seqfasta=$opt_v;                   }
    if($opt_x){ $seqdat  =$opt_x;                   }
    if($opt_y){ $tpldat  =$opt_y;                   }
    if($opt_z){ $cntdat  =$opt_z;                   }
    if($opt_o){ $outd    =addAbsPath($opt_o); $f++; }
    if($f<2){ print "Hello\n";  &message(); }
    if($opt_h){ &message(); }
}

print "...running mkdat.pl...\n";

########### setup  the environment and Working DIRectory ###
$ENV{'PATH'}="/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/pgi/linux86/bin";
$ENV{'LD_LIBRARY_PATH'}="/usr/local/lib:/usr/lib:/lib";

%ts=(
     'GLY'=>'G',
     'ALA'=>'A',
     'VAL'=>'V',
     'LEU'=>'L',
     'ILE'=>'I',
     'SER'=>'S',
     'THR'=>'T',
     'CYS'=>'C',
     'MET'=>'M',
     'PRO'=>'P',
     'ASP'=>'D',
     'ASN'=>'N',
     'GLU'=>'E',
     'GLN'=>'Q',
     'LYS'=>'K',
     'ARG'=>'R',
     'HIS'=>'H',
     'PHE'=>'F',
     'TYR'=>'Y',
     'TRP'=>'W',

     'ASX'=>'B',
     'GLX'=>'Z',
     'UNK'=>'X',

     'G'=>'GLY',
     'A'=>'ALA',
     'V'=>'VAL',
     'L'=>'LEU',
     'I'=>'ILE',
     'S'=>'SER',
     'T'=>'THR',
     'C'=>'CYS',
     'M'=>'MET',
     'P'=>'PRO',
     'D'=>'ASP',
     'N'=>'ASN',
     'E'=>'GLU',
     'Q'=>'GLN',
     'K'=>'LYS',
     'R'=>'ARG',
     'H'=>'HIS',
     'F'=>'PHE',
     'Y'=>'TYR',
     'W'=>'TRP',

     'a'=>'CYS',
     'b'=>'CYS',
     'c'=>'CYS',
     'd'=>'CYS',
     'e'=>'CYS',
     'f'=>'CYS',
     'g'=>'CYS',
     'h'=>'CYS',
     'i'=>'CYS',
     'j'=>'CYS',
     'k'=>'CYS',
     'l'=>'CYS',
     'm'=>'CYS',
     'n'=>'CYS',
     'o'=>'CYS',
     'p'=>'CYS',
     'q'=>'CYS',
     'r'=>'CYS',
     's'=>'CYS',
     't'=>'CYS',
     'u'=>'CYS',
     'v'=>'CYS',
     'w'=>'CYS',
     'x'=>'CYS',
     'y'=>'CYS',
     'z'=>'CYS',

     'B'=>'ASX',
     'Z'=>'GLX',
     'X'=>'CYS',
     );

@ww=qw(
       0.05
       0.10
       0.15
       0.20
       0.25
       0.30
       0.35
       0.40
       0.45
       0.50
       0.55
       0.60
       0.65
       0.70
       0.75
       0.80
       0.85
       );

@AA=qw(
       C
       M
       F
       I
       L
       V
       W
       Y
       A
       G
       T
       S
       Q
       N
       E
       D
       H
       R
       K
       P
       );

################# default directories and files #############################
$data_dir=addAbsPath(".");
$bin_dir="$codedir/perl/tasser/tasser1.0/mkinput/bin";
# comm_dir for *.comm
$comm_dir="$codedir/perl/tasser/tasser1.0/mkinput/data";
$work_dir=addAbsPath("./tmp_mkdat");
$libdir="/library/yzhang/PDB/";
$blastdir="/library/yzhang/blast";
$db="/library/yzhang/nr/nr.filter";
$seqfasta='seq.fasta';
$seqdat='seq.dat';
$tpldat='template.dat';
$cntdat='contact.dat';
# customary files and directories
&parse_input();
$libdir.='/' unless $libdir=~/\/$/ ; #libdir must end in '/'

if($seqfasta=~/\// || $seqdat=~/\// || $tpldat=~/\// || $cntdat=~/\//){
    print "ERROR flags v,x,y,z don't support '/'\n";
}
################ working directory ########################
`/bin/mkdir -p $work_dir`;
chdir "$work_dir"; 
unless($work_dir=~/^\//){ chomp($work_dir=`pwd`); }
#`/bin/rm -f $work_dir/*`;
`/bin/cp $data_dir/$seqfasta ./seq.fasta`;
`/bin/cp $data_dir/$seqdat ./seq.dat`;
`/bin/cp $data_dir/$tpldat ./template.dat`;
`/bin/cp $data_dir/$cntdat ./contact.dat`;
`/bin/cp $comm_dir/wgt.tar.bz2 .`;
`/usr/bin/bunzip2 wgt.tar.bz2`;
`/bin/tar -xvf wgt.tar`;

`/bin/cp $bin_dir/zal ./zal`;
`/bin/cp $bin_dir/dat ./dat`;
`/bin/cp $bin_dir/solve ./solve`;

################ make 'protein.seq' #################
@seqfastas=`cat seq.fasta`;
$sequence="";
foreach $seqfasta(@seqfastas){
    goto pos6 if($seqfasta=~/\>/);
    $seqfasta=~s/\s//mg;
    $seqfasta=~s/\n//mg;
    $sequence=$sequence.$seqfasta;
  pos6:;
}
$Lch=length $sequence;
open(seq,">protein.seq");
printf seq ">protein\n";
for($i=1;$i<=$Lch;$i++){
    $a=substr($sequence,$i-1,1);
    printf seq "$a";
    $seqQ{$i}=$a;
    $log{$i,$seqQ{$i}}++;
    if($i==int($i/60)*60){
	printf seq "\n";
    }
}
printf seq "\n";
close(seq);

############## make exp.dat #########################
print "...mkdat.pl calling blastpgp...\n...running blastpgp...\n"; 
system("$blastdir/blastpgp -i protein.seq -Q protein.mat3 -d $db -o protein.out -e 1e-3 -h 1e-3 -M BLOSUM62 -j 3 -m 6");
print "...finished blastpgp...\n";

print "...mkdat.pl calling solve...\n...runing solve...\n";
foreach $w(@ww){
    `./solve weight.$w protein`; #protein.neu
    open(neu,"protein.neu");
    <neu>=~/(\d+)/;
    $Lch=$1;
    for($i=1;$i<=$Lch;$i++){
	<neu>=~/(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)/;
	$exp{$i,$w}=$5;
    }
    close(neu);
}
print "...finished solve...\n";
open(exp1,">exp.dat");
printf exp1 "$Lch  ";
foreach $w(@ww){
    printf exp1 " %5s",$w;
}
printf exp1 "\n";
for($i=1;$i<=$Lch;$i++){
    printf exp1 "%5d",$i;
    foreach $w(@ww){
	printf exp1 " %5d",$exp{$i,$w};
    }
    printf exp1 "\n";
}
close(exp1);

########### run psi-blast #######################
print "...mkdat.pl calling blastpgp...\n...running blastpgp...\n"; 
system("$blastdir/blastpgp  -b 1000 -j 3 -h 0.001 -d $db -i protein.seq -C psitmp.chk > blast.out");
print "...finished blastpgp...\n";

########### calculate 'protein.frq' ###################
open(blast,"blast.out");
while($line=<blast>){
    if($line=~/Results from round\s+(\d+)/){
	$ROUND=$1;
    }
}
seek(blast,0,0);
$it=0;
while($line=<blast>){
    if($line=~/round\s+$ROUND/){
	while($line=<blast>){
	    if($line=~/Expect =\s*(\S+)/){
		$ev=$1;
		<blast>=~/Identities =\s*\S+\s+\((\S+)\%/;
		<blast>;
		$id=$1;
		$it++;
		if($ev<0.001 && $id < 98){
		    while($line=<blast>){
			if($line=~/Query\:\s*(\d+)\s+(\S+)\s+(\S+)/){
			    $i1=$1;
			    $seq1=$2;
			    <blast>;
			    <blast>=~/Sbjct\:\s*(\S+)\s+(\S+)\s+(\S+)/;
			    $seq2=$2;
			    <blast>;
			    ###
			    $L=length $seq1;
			    $m1=$i1-1;
			    for($i=1;$i<=$L;$i++){
				$q1=substr($seq1,$i-1,1);
				$q2=substr($seq2,$i-1,1);
				$m1++ if($q1 ne '-');
				if($q1 ne '-' && $q2 ne '-'){
				    $log{$m1,$q2}++;
				}
			    }
			    ###
			}else{
			    goto pos1;
			}
		    }
		}
	      pos1:;
	    }
	}
    }
}
close(blast);
open(freq,">protein.frq");
printf freq "$Lch\n";
for($i=1;$i<=$Lch;$i++){
    printf freq "%3d",$i;
    $norm=0;
    foreach $A(@AA){
	$norm+=$log{$i,$A};
    }
    foreach $A(@AA){
	printf freq "%10.7f",$log{$i,$A}/$norm;
    }
    printf freq "\n";
}
close(freq);

########### make in.dd #############
open(in,">in.dd");
printf in "protein.seq\n";
printf in "protein.frq\n";
printf in "$libdir\n";
printf in "seq.pdb\n";
printf in "seq.dat\n";
printf in "0.3\n"; #sequence cutoff
printf in "template.dat\n";
printf in "contact.dat\n";
printf in "8   15.5\n"; #zscore cutoff for medium and easy classes for PROSPECTOR_3 templates. This two numbers are adjustable when PROSPECTOR_3 is updated.
close(in);

########### run dat ################
print "...mkdat.pl calling zal...\n...running zal...\n"; 
`./zal`;
print "...finished zal...\n";

print "...mkdat.pl calling dat...\n...running dat...\n"; 
`./dat`;
print "...finished dat...\n";

`/bin/mkdir -p $outd`;

@listOut=("comb.dat","combCA.dat","dist.dat","distL.dat","chain.dat","par.dat","exp.dat");
for $file (@listOut){
    if(system("ls $file >/dev/null 2>/dev/null")){
	print "Error: file $file does not exists\n";
	exit(1);
    }
    `/bin/cp $file  $outd`;
}

#`/bin/rm -rf $work_dir`;
`sync`;
`sync`;
sleep(2);
print "...finished mkdat.pl...\n";

exit(0);

File: python/tasser/tasser1.0/mkinput/mkdat.py
#!/usr/bin/python
import sys,os,re,copy
from inputArgs.inputArgs import inpHand,addAbsPath
from utilities.codedir import codedir,libseq,yangpdb
from utilities.small_utilities import junkName,Bye
from jobs.job import pastry
from seq.fastaManager import importFastaEntry,writeFasta
from seq.letters import valid_A
from jobs.job import pastry
from seq.blastManager import blastRun,blastAlignment

zal=codedir+'/perl/tasser/tasser1.0/mkinput/bin/zal'
dat=codedir+'/perl/tasser/tasser1.0/mkinput/bin/dat'

def formatTempl(before,after):
    p0=re.compile('TER')
    p1=re.compile('ATOM')
    inf=open(before,'r')
    outf=open(after,'w')
    n=os.popen('grep TER '+before+'|wc|tr -s " "|cut -d" " -f2').readline()
    outf.write(n)
    c=1 #template number
    for line in inf.readlines():
        if p0.search(line):
            outf.write('TER\n')
            c+=1 #another template read
        else:
            if p1.search(line):
                aa=line[17:20]
                n1=int(line[22:26]) ; n2=int(line[55:59])
                rx=float(line[30:38]) ; ry=float(line[38:46]) ; rz=float(line[46:54])
                outf.write('ATOM %6d  CA  %s%6d    %8.3f%8.3f%8.3f%5d\n'%(n1,aa,n1,rx,ry,rz,n2))
            else:
                it=line.split()
                it[1]=int(it[1])  ;  it[2]=float(it[2])
                outf.write('%6d%12.5f%7d     %s\n' % (it[1],it[2],c,it[0]) )
    inf.close()
    outf.close()
    
def formatCont(before,after):
    inf=open(before,'r')
    outf=open(after,'w')
    outf.write(inf.readline()) #fist line
    for line in inf.readlines():
        it=line.split()
        it[0]=int(it[0])  ;  it[1]=int(it[1])
        outf.write( '%5d %5d\n'%(it[0],it[1]) )
    inf.close()
    outf.close()

#gen_dat expects to find the following files: contact.dat, template.dat, seq.dat, seq.fasta
def gen_dat(libdir=yangpdb,blastf='',outd='.',co=0.3):
    outd=addAbsPath(outd)
    #write protein.seq file from seq.fasta
    header,seq=importFastaEntry(open('seq.fasta','r')) #;Bye(header+'\n'+seq)
    writeFasta('protein',seq,filename='protein.seq',n=60)

    #create blast.out
    if blastf: os.system('/bin/cp '+blastf+' blast.out')
    else:
        blastOuts=blastRun('protein.seq') #run psi-blast
        os.system('/bin/cp '+blastOuts['blast']+' blast.out')

    #create protein.frq
    prof={}
    for A in valid_A: prof[A]=0 #initialize generic profile
    log=[]
    for i in range(0,len(seq)): log.append(copy.deepcopy(prof))    
    #go to last round
    fpt=open('blast.out','r') ; line=fpt.readline()
    r=re.compile('\s*Results from round 3|CONVERGED\!')
    while not r.match(line):    line=fpt.readline() #scan until match
    #read all blast alignments
    while(1):
        alg=blastAlignment(fpt,querySeq=seq) #extended alignment to full sequence of "Query"
        if alg.error: break #reading error or just end-of-file reached
        if alg.pd['I']<0.98 and alg.pd['E']<0.001:
            gs=alg.Yalg.project( alg.ix['Sbjct'], alg.ix['Query'] ).gs #gapped sequence of Sbjct onto Query
            for i in range(0,len(gs)):
                if gs[i] in valid_A: log[i][ gs[i] ]+=1
    pout=open('protein.frq','w') ; pout.write(`len(seq)`+'\n')
    for i in range(0,len(seq)):
        pout.write('%3d'%(i+1))
        norm=0 ; prof=log[i]
        for A in valid_A: norm+=prof[A]
        for A in 'CMFILVWYAGTSQNEDHRKP':
            x=1.0*prof[A]
            if norm!=0: x=x/norm
            pout.write('%10.7lf'%x) #precise A order required
        pout.write('\n')
    pout.close()

    #write in.dd
    pout=open('in.dd','w')
    pout.write('protein.seq\nprotein.frq\n')
    pout.write(libdir+'/\n') #notice the trailing '/'
    pout.write('seq.pdb\nseq.dat\n')
    pout.write('%5.3lf'%(co)+'\n') #sequence cutoff
    pout.write('template.dat\ncontact.dat\n')
    pout.write('8   15.5\n') #zscore cutoff for medium and easy classes for PROSPECTOR_3 templates. This two numbers are adjustable when PROSPECTOR_3 is updated
    pout.close()

    #run black box
    os.system('/bin/cp '+zal+' '+dat+' .')
    print 'running zal'  ;  os.system('./zal')
    print 'running dat'  ;  os.system('./dat')
    for file in ['comb.dat','combCA.dat','dist.dat','distL.dat','chain.dat']:
        if not os.path.exists('./'+file): Bye('From mkdat.py: failed to generate '+file)
    if os.getcwd()!=outd: pastry('/bin/mv comb*.dat dist*.dat chain.dat '+outd)
    
#execute as standalone script
if __name__=='__main__':
    ih=inpHand('Usage: mkdat.py',
               ' -a _R_header five letter PDB header',
               ' -b _AR_fastaf sequence file in FASTA format',
               ' -c _AR_seqdat secondary structure propensity',
               ' -d _A_templ template.dat-like file',
               ' -e _A_templ2 template file from prospector (in place of -d flag)',
               ' -f _A_pred contact.dat-like file',
               ' -g _A_pred2 predicted contacts from prospector (in place of -d flag)',
               ' -i _A_libdir library of structures (def:/library/yzhang/PDB)',
               ' -k _A_outd output directory (def: current dir)',
               ' -l _A_blastf results from a previous blast run',
               ' -m __co sequence cutoff(def:0.3)'
               )
    ih.parse(locals(),sys.argv)

    if not co:co=0.3
    else: co=float(co)

    print '...running mkdat.py...'
    if not templ and not templ2: ih.abort('ERROR: pass either -d or -e flag')
    if not pred and not pred2  : ih.abort('ERROR: pass either -f or -g flag')
    if not libdir: libdir=yangpdb
    if not outd: outd=os.getcwd()
    for file in (fastaf,seqdat,templ,templ2,pred,pred2,libdir,blastf):
        if file and not os.path.exists(file):ih.abort('ERROR: '+file+' does not exists')

    if not os.path.exists('./seq.fasta'): os.system('/bin/cp '+fastaf+' ./seq.fasta')
    if not os.path.exists('./seq.dat'): os.system('/bin/cp '+seqdat+' ./seq.dat')

    #template file
    if templ2 and not templ:
        templ='./template.dat'  ;  formatTempl(templ2,templ)
    else:
        if not os.path.exists('./template.dat'):
            os.system('/bin/cp '+templ+' ./template.dat 2>/dev/null')

    #contacts file
    if pred2 and not pred:
        pred='./contact.dat'  ;  formatCont(pred2,pred)
    else:
        os.system('/bin/cp '+pred+' ./contact.dat 2>/dev/null')

    gen_dat(libdir=libdir,blastf=blastf,outd=outd,co=co)
    
    #print '...mkdat.py calling mkdat.pl...'
    #pastry(codedir+'/perl/tasser/tasser1.0/mkinput/mkdat.pl -l '+libdir+' -d . -o '+outd)
        
    print '...finished mkdat.py...'
    sys.exit(0)
